clang/labs/lab1/durandin_vladimir/AlwaysInlineAttributePlugin.cpp

```diff
@@ -0,0 +1,68 @@
+#include "clang/AST/ASTConsumer.h"
+#include "clang/AST/RecursiveASTVisitor.h"
+#include "clang/Frontend/CompilerInstance.h"
+#include "clang/Frontend/FrontendPluginRegistry.h"
+
+class AddAlwaysInlineConsumer : public clang::ASTConsumer {
+public:
+  bool hasCondition(clang::Stmt *S) {
+    if (!S) {
+      return false;
+    }
+
+    if (llvm::isa<clang::IfStmt>(S) || llvm::isa<clang::SwitchStmt>(S) ||
+        llvm::isa<clang::WhileStmt>(S) || llvm::isa<clang::DoStmt>(S) ||
+        llvm::isa<clang::ForStmt>(S)) {
+      return true;
+    }
+
+    for (clang::Stmt *Child : S->children()) {
+      if (hasCondition(Child)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  bool HandleTopLevelDecl(clang::DeclGroupRef D) override {
+    clang::FunctionDecl *FD = nullptr;
+
+    for (clang::Decl *Decl : D) {
+      FD = clang::dyn_cast<clang::FunctionDecl>(Decl);
+      if (FD) {
+        if (!hasCondition(FD->getBody())) {
+          FD->addAttr(
+              clang::AlwaysInlineAttr::CreateImplicit(FD->getASTContext()));
+          FD->print(llvm::outs() << "__attribute__((always_inline)) ");
+        } else
+          FD->print(llvm::outs());
```

Comment:
In my implementation, these lines are not useless, since the tests use the standard output stream (`std::out`) and are redirected to FileCheck to check the work.

I tried to make it so that I could call a dump when the script was launched (`clang -cc1 -ast-dump ...`), but nothing was displayed on my `WSL`, no matter how hard I tried (I even built the `clang-ast-dump` target separately). Therefore, I simply don’t know how to check correctly in tests using the AST tree, without **EXPLICITLY** calling the `llvm::outs()` or `->dump()` functions.